fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/QueryPlan.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * QueryPlan.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.plans;
+
+import com.apple.foundationdb.API;
+import com.apple.foundationdb.record.EvaluationContext;
+import com.apple.foundationdb.record.ExecuteProperties;
+import com.apple.foundationdb.record.PlanHashable;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.provider.common.StoreTimer;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordStore;
+import com.apple.foundationdb.record.query.plan.temp.expressions.RelationalPlannerExpression;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * An executable query plan. A query plan can be executed against a record store to get a stream of items.
+ * Unlike a {@link RecordQueryPlan} (which extends this interface), implementations are not required to
+ * return records but might instead return items of an arbitrary type. Note that
+ * {@link com.apple.foundationdb.record.query.plan.QueryPlanner QueryPlanner}s always return {@code RecordQueryPlan}s,
+ * but the query plan tree might contain intermediate plans that return something other than records.
+ *
+ * @param <T> the type of element produced by executing this plan
+ */
+@API(API.Status.EXPERIMENTAL)
+public interface QueryPlan<T> extends PlanHashable, RelationalPlannerExpression {
+
+    /**
+     * Execute this query plan.
+     * @param store record store from which to fetch records
+     * @param context evaluation context containing parameter bindings
+     * @param continuation continuation from a previous execution of this same plan
+     * @param executeProperties limits on execution
+     * @return a cursor of records that match the query criteria
```

Comment:
I'm not sure that this method necessarily produces a cursor of "records". Perhaps it's a "cursor of `<T>`s" or something like that? It might be more natural if `T` gets a pronounceable name.

Reply:
I went with "items", here and elsewhere. Maybe that's not the best name?